icu4c/source/common/locid.cpp

```diff
@@ -1457,7 +1460,8 @@ bool
 AliasReplacer::replace(const Locale& locale, CharString& out, UErrorCode status)
 {
     data = AliasData::singleton(status);
-    if (U_FAILURE(status)) {
+    if (U_FAILURE(status) || data == nullptr) {
+        status = U_MEMORY_ALLOCATION_ERROR;
```

Comment:
Don't clobber an existing failure.
Better:
```
if (U_FAILURE(status)) {
    return false;
}
if (data == nullptr) {
    status = U_MEMORY_ALLOCATION_ERROR;
    return false;
}
```

Alternatively, and a bit cleaner I think, would be to check for nullptr after "new" at the end of the singleton() function, and set the memory error there. Then you need not change this call site. Basically, singleton() should either return with "success" and a non-nullptr, or a failure and nullptr.

Reply:
PS: Before `return new AliasData(` please check for failure and return nullptr.

Reply:
So... I realize that singleton() is not the right place. The "new" is in `AliasDataBuilder::build()`. And once that behaves as it should, you should then remove the `if (gSingleton == nullptr) {` check from `AliasData::loadData()` so that another error code does not get clobbered.

Sorry for not paying more attention during my earlier reviews.

Reply:
I somehow missed a & in the loadData and cause issue and didn't realize till now.
Also we should return earlier if the status has already error so we won't call the init once code with status is error because that code will not be run again. 
We also need to set the error after the line of calling the init once call and set the status to memory error because the loadData will not be called the second time. 

PTAL